### PR TITLE
Preview commit

### PR DESCRIPTION
**Pull request by grosser**

### CODE REVIEW

File: app/models/docker_builder_service.rb [hunk 1]
```diff
@@ -21,9 +22,32 @@ def run!(image_name: nil, push: false)
...
 
-      if build_image(tmp_dir)
-        push_image(image_name) if push
-        build.docker_image.remove(force: true)
+      if kuber_job
+        kuber_job = Kubernetes::JobExecutor.new(output, job: job)
+        docker_ref = image_name.presence || build.label.try(:parameterize).presence || 'latest'
+
+        success, build_log = kuber_job.build!(build.git_sha,
+                                              repo: project.docker_repo, tag: docker_ref, push: push,
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: please use 2-space indent, uses less space and does not need to be re-indented on every change

File: app/models/docker_builder_service.rb [hunk 2]
```diff
@@ -21,9 +22,32 @@ def run!(image_name: nil, push: false)
...
+                                              registry: registry_credentials)
+        if push
+          build.docker_ref = docker_ref
+          build.docker_repo_digest = nil
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: why does it unset this ?

File: app/models/docker_builder_service.rb [hunk 3]
```diff
@@ -21,9 +22,32 @@ def run!(image_name: nil, push: false)
...
+        end
+
+        if success and !build_log.blank?
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: prefer `&&` and `!blank -> present`

File: app/models/docker_builder_service.rb [hunk 4]
```diff
@@ -21,9 +22,32 @@ def run!(image_name: nil, push: false)
...
+          build_log.each_line do |line|
+            if (matches = DOCKER_REPO_REGEX.match(line))
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: FYI a bit shorter and no match object: `if (match = line[DOCKER_REPO_REGEX, 1]`

File: app/models/docker_builder_service.rb (hunk 3 above)
Comment (grosser) [inline app/models/docker_builder_service.rb]: is success and no log a valid usecase ? ... seems weird ...

File: app/models/docker_builder_service.rb (hunk 4 above)
Comment (grosser) [inline app/models/docker_builder_service.rb]: should blow up if it cannot find the ref too ?

File: plugins/kubernetes/app/models/build_job.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+---
```
Comment (grosser) [inline plugins/kubernetes/app/models/build_job.yml]: a yml file is not a model ... can put that somewhere else like config ?

File: plugins/kubernetes/app/models/build_job.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      restartPolicy: Never
+      nodeSelector:
+        role: build-server
+      volumes:
+        - name: cached-repo
+          hostPath:
+           path: /opt/cached_repos
+        - name: docker-sock
+          hostPath:
+           path: /var/run/docker.sock
```
Comment (grosser) [inline plugins/kubernetes/app/models/build_job.yml]: lot of hardcoding here ... maybe generate that file ?

File: plugins/kubernetes/app/models/kubernetes/api/job.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      def complete?
+        if @job.status.conditions.present?
+          complete = @job.status.conditions.find { |c| c['type'] == 'Complete'}
+          complete && complete['status'] == 'True'
+        else
+          false
+        end
+      end
+
+      def labels
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/api/job.rb]: `private` ?

File: plugins/kubernetes/app/models/kubernetes/api/job.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      def namespace
+        @job.metadata.try(:namespace)
+      end
+
+      def failure?
+        @job.status == "Failure"
+      end
+
+      def complete?
+        if @job.status.conditions.present?
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/api/job.rb]: FYI
```Ruby
def complete?
  (@job.status.conditions || []).detect { |c| break c['status'] == 'True' if c['type'] == 'Complete' }
end
```

File: plugins/kubernetes/app/models/kubernetes/job_config_file.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  class JobConfigFile
+    attr_reader :file_path, :job
+
+    def initialize(contents, file_path)
+      @file_path = file_path
+      @config_file = Kubernetes::Util.parse_file(contents, file_path)
+      parse_file
+    end
+
+    def parse_file
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/job_config_file.rb]: private ?

File: plugins/kubernetes/app/models/kubernetes/job_config_file.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+# - Base Kubernetes config file ???
+# - Base YAML Kubernetes resource definition ??
+
+module Kubernetes
+  class JobConfigFile
+    attr_reader :file_path, :job
+
+    def initialize(contents, file_path)
+      @file_path = file_path
+      @config_file = Kubernetes::Util.parse_file(contents, file_path)
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/job_config_file.rb]: it's not a file ... so call it `@config` ?

File: plugins/kubernetes/app/models/kubernetes/job_config_file.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+      hash = Array.wrap(@config_file).detect { |doc| doc['kind'] == type }.freeze
+      hash.dup.with_indifferent_access unless hash.nil?
+    end
+
+    #
+    # Inner class
+    #
+    class Job < RecursiveOpenStruct
+      def initialize(hash = nil, args = {})
+        args.merge!(recurse_over_arrays: true)
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/job_config_file.rb]: sweet ... I needed that option a few times already :D

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        return false, nil if start + WAIT_FOR_JOB < Time.now
+        return true, job if job.complete?
+
+        sleep TICK
+      end
+    end
+
+    def build!(reference, repo:, tag:, push: false, registry:)
+      @output.puts "### Running a remote Docker build job"
+      File.open(build_job_config_path, 'r') do |contents|
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: can do a `content = File.read` to have 1 less block around everything

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        kuber_job = Kubernetes::JobConfigFile.new(contents, build_job_config_path).job
+        kuber_job.first_container.args = [@job.project.repository_directory, reference, repo, tag, push ? "yes" : "no"]
+        kuber_job.first_container.env = [
+          {name: 'DOCKER_REGISTRY_USER', value: registry["username"] || ""},
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: maybe fail if any of these are missing ?

File: plugins/kubernetes/config/initializers/watchers.rb [hunk 14]
```diff
@@ -6,7 +6,7 @@ module Kubeclient
   class Client
     # Add Deployment and Daemonset waiting for PR:
     # https://github.com/abonas/kubeclient/pull/143
-    NEW_ENTITY_TYPES = %w(Deployment DaemonSet).map do |et|
+    NEW_ENTITY_TYPES = %w(Deployment DaemonSet Job).map do |et|
```
Comment (grosser) [inline plugins/kubernetes/config/initializers/watchers.rb]: `@jonmoter` had a patch to get this all into kubeclient / PR might need a rebase

**Comment by hdt91** [comment]
@grosser I created a new PR (https://github.com/zendesk/samson/pull/1076) and addressed most of your comments in this old PR. Closing it now. Thanks for the feedback :D

**Pull request closed without merging**